client/allocrunner/group_service_hook_test.go

```diff
@@ -293,3 +294,161 @@ func TestGroupServiceHook_getWorkloadServices(t *testing.T) {
 	services := h.getWorkloadServicesLocked()
 	must.Len(t, 1, services.Services)
 }
+
+func TestGroupServiceHook_PreKill(t *testing.T) {
+	ci.Parallel(t)
+	logger := testlog.HCLogger(t)
+
+	t.Run("waits for shutdown delay", func(t *testing.T) {
+		alloc := mock.Alloc()
+		alloc.Job.TaskGroups[0].Networks = []*structs.NetworkResource{}
+		tg := alloc.Job.LookupTaskGroup(alloc.TaskGroup)
+		tg.Services = []*structs.Service{
+			{
+				Name:      "testconnect",
+				PortLabel: "9999",
+				Connect: &structs.ConsulConnect{
+					SidecarService: &structs.ConsulSidecarService{},
+				},
+			},
+		}
+		delay := 200 * time.Millisecond
+		tg.ShutdownDelay = &delay
+
+		consulMockClient := regMock.NewServiceRegistrationHandler(logger)
+
+		regWrapper := wrapper.NewHandlerWrapper(
+			logger,
+			consulMockClient,
+			regMock.NewServiceRegistrationHandler(logger))
+
+		shutDownCtx, cancel := context.WithTimeout(context.Background(), delay*2)
+		defer cancel()
+
+		h := newGroupServiceHook(groupServiceHookConfig{
+			alloc:             alloc,
+			serviceRegWrapper: regWrapper,
+			shutdownDelayCtx:  shutDownCtx,
+			restarter:         agentconsul.NoopRestarter(),
+			taskEnvBuilder:    taskenv.NewBuilder(mock.Node(), alloc, nil, alloc.Job.Region),
+			logger:            logger,
+			hookResources:     cstructs.NewAllocHookResources(),
+		})
+
+		successChan := make(chan struct{}, 1)
+		go func() {
+			before := time.Now()
+			h.PreKill()
+			after := time.Now()
+			// we respected the shutdown_delay and did not return immediately
+			if after.Sub(before) < delay {
+				t.Fail()
+			}
```

Comment:
The stdlib has a nice little helper function for this. And I think the comment is backwards?

```suggestion
			// we did not respect the shutdown_delay and returned immediately
			if time.Since(before) < delay {
				t.Fail()
			}
```